elyra/util/operation_parser.py

```diff
@@ -111,53 +99,89 @@ def __init__(self, operation_filepath):
     def type(self):
         raise NotImplementedError()
 
+    @property
+    def model(self):
+        return self.build_model()
+
     @property
     def operation_filepath(self):
         # TODO : Ensure this is getting the abs path
         return self._operation_filepath
 
+    @property
+    @abstractmethod
+    def regex_dict(self) -> Dict[str, List]:
+        # a dictionary of matching regex expressions to use keyed to the json stanza
+        # Parsing for ENVs e.g. regex_dict['NAME_OF_JSON_STANZA'] = ['REGEX_EXP_1', 'REGEX_EXP_2']
+        raise NotImplementedError()
+
     @abstractmethod
-    def env_list(self):
+    def file_as_code_chunks(self) -> List[str]:
+        # A list of code `chunks` to look through. The size and length of the `chunks` is determined
+        # usually by the language. For now the concrete implementations below is by line-by-line in file
         raise NotImplementedError()
 
+    def build_model(self):
+        """
+        Builds a model dictionary of all the regex matches it for each key in the regex dictionary
+        :return:
+        """
+        model = dict()
+        for chunk in self.file_as_code_chunks():
+            for key in self.regex_dict.keys():
+                regex_list = self.regex_dict[key]
+                for pattern in regex_list:
+                    match_found = re.search(pattern, chunk)
+                    if match_found and match_found.group(1) not in model[key]:
+                        model[key].append(match_found.group(1))
+
+        return model
+
+
+class NotebookOperationParser(OperationParser):
+
+    _type = "notebook"
+
+    def __init__(self, operation_filepath):
+        super().__init__(operation_filepath)
+        self.notebook = nbformat.read(open(self.operation_filepath), as_version=4)
+        self.language = self.notebook['metadata']['kernelspec']['language']
+
+    def type(self):
+        return self._type
+
+    def file_as_code_chunks(self) -> List[str]:
+        for cell in self.notebook.cells:
+            if cell.source:
+                file_by_line = cell.source.split('\n')
+                return file_by_line
+
+    def regex_dict(self) -> Dict[str, List]:
+        if self.language == 'python':
+            return PythonOperationParser(self.operation_filepath).regex_dict()
+        elif self.language == 'r':
+            pass
+            # return ROperationParser.regex_dict()
+
 
 class PythonOperationParser(OperationParser):
 
     _type = "python"
 
     def __init__(self, operation_filepath):
         super().__init__(operation_filepath)
+        self.python_file = open(self.operation_filepath)
```

Comment:
This file descriptor will be leaked.  Per my earlier comment, it might be better to have the operation parses use a Context Manager (i.e., `with` clause) that gathers some set of lines and feeds them to the contained language parser.